### PR TITLE
Fixing byte-order consistency/mismatch for cross-endian platform

### PR DESCRIPTION
**Pull request by pradghos**
- Addressing https://github.com/joblib/joblib/issues/1123

### CODE REVIEW

**Comment by pradghos** [comment]
Circle CI failed at ./continuous_integration/build_doc.sh

  ```
  result = self._call_chain(self.handle_open, protocol, protocol +
  File "/home/circleci/miniconda/envs/testenv/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/home/circleci/miniconda/envs/testenv/lib/python3.9/urllib/request.py", line 1389, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/home/circleci/miniconda/envs/testenv/lib/python3.9/urllib/request.py", line 1349, in do_open
    raise URLError(err)
  urllib.error.URLError: <urlopen error [Errno 110] Connection timed out>

  Makefile:11: recipe for target 'html' failed
```

Before this commit `https://github.com/joblib/joblib/pull/1181/commits/d9e5adbcc5478159310ebd58b3c070c5feea7e7d` all were green except `ci/circleci: build` 

Now `test_nested_loop_error_in_grandchild_resource_tracker_silent` failed which looks like unrelated to this PR.

Any pointers would really help . Thank you !

**Comment by pradghos** [comment]
@ogrisel It would be great if you can take a look. Thanks in advance !

**Comment by lesteve** [comment]
Thanks a lot for this PR! What would be great is to have a test for this.

Is there a way to store a pickle string obtained from a big endian machine and test that it can be loaded fine in the CI (little endian machine)?

I think we can ignore the CI failures as they do seem spurious.

**Comment by ogrisel** [comment]
Alternatively could you just please add a couple of unit tests for `_is_numpy_array_byte_order_mismatch` alone (independently of `jolib.load` / `dump`)?

**Comment by pradghos** [comment]
@lesteve @ogrisel Thank you very much for the review. I have added test. Please let me know in case of any further comments.

File: joblib/test/test_numpy_pickle.py [hunk 1]
```diff
@@ -457,6 +463,35 @@ def test_joblib_pickle_across_python_versions():
...
+                          dtype=[('', '>i8'), ('', '>f8')]),
+                 np.arange(3, dtype=np.dtype('>i8')),
+                 np.arange(3, dtype=np.dtype('>f8'))]
+
+    # Verify the byteorder mismatch is correctly detected.
+    for array in be_arrays:
+        if sys.byteorder == 'big':
+            assert _is_numpy_array_byte_order_mismatch(array) is False
+        else:
+            assert _is_numpy_array_byte_order_mismatch(array) is True
```
Comment (ogrisel) [inline joblib/test/test_numpy_pickle.py:479]: style nitpick:

```suggestion
            assert not _is_numpy_array_byte_order_mismatch(array)
        else:
            assert _is_numpy_array_byte_order_mismatch(array)
```

File: joblib/test/test_numpy_pickle.py [hunk 2]
```diff
@@ -457,6 +463,35 @@ def test_joblib_pickle_across_python_versions():
...
+                          dtype=[('', '<i8'), ('', '<f8')]),
+                 np.arange(3, dtype=np.dtype('<i8')),
+                 np.arange(3, dtype=np.dtype('<f8'))]
+
+    # Verify the byteorder mismatch is correctly detected.
+    for array in le_arrays:
+        if sys.byteorder == 'little':
+            assert _is_numpy_array_byte_order_mismatch(array) is False
+        else:
+            assert _is_numpy_array_byte_order_mismatch(array) is True
```
Comment (ogrisel) [inline joblib/test/test_numpy_pickle.py:492]: style nitpick

```suggestion
    for array in le_arrays:
        if sys.byteorder == 'little':
            assert not _is_numpy_array_byte_order_mismatch(array)
        else:
            assert _is_numpy_array_byte_order_mismatch(array)
```

**Review by ogrisel — COMMENTED** [review]
Some nitpick. I plan to push a new commit to better test the byteswap operation itself after that.

**Review by ogrisel — APPROVED** [review]

**Comment by ogrisel** [comment]
Merged!, thanks @pradghos!

**Pull request merged**